Add docs for Kafka source and Kafka target (#1011)
triggermesh/triggermesh: docs/targets/kafka.md

# Kafka event target

> This event target integrates with Kafka, using received Cloud Event messages to publish a message to a Kafka Cluster.

Reviewer: ```suggestion
This event target integrates with Kafka, where any Cloud Event received are published to a Kafka Cluster.
```

## Contents

- [Kafka event target](#kafka-event-target)

>   - [Contents](#contents)

Reviewer: is this a typo?

  - [Prerequisites](#prerequisites)
  - [Creating a Kafka Targe with SASL-PLAIN](#creating-a-kafka-target-with-sasl-plain)

>   - [Creating a Kafka Targe with Kerberos-SSL](#creating-a-kafka-target-with-kerberos-ssl)

Reviewer: same suggestion as the other doc, i.e. create sub-sections

    - [Status](#status)
    - [Kafka Target as an event Sink](#kafka-target-as-an-event-sink)
    - [Sending messages to the Kafka Target](#sending-messages-to-the-kafka-target)

## Prerequisites

A running Kafka cluster.

## Creating a Kafka Target with SASL-PLAIN

> 

Reviewer: some description...

logs and events can provide detailed information about the process. A Status
summary is added to the KafkaTarget object informing of the all conditions
that the target needs.

When ready the `status.address.url` will point to the internal point where Cloud Events should be sent.

### Kafka Target as an event Sink

A Kafka Target is addressable, which means it can be used as a Sink for Knative components.

```yaml
apiVersion: eventing.knative.dev/v1beta1
kind: Trigger
metadata:
  name: kafka-sample-trigger
spec:
  broker: default
  subscriber:
    ref:
      apiVersion: targets.triggermesh.io/v1alpha1
      kind: KafkaTarget
      name: triggermesh-kafka
```

- A sample sink binding to a Kafka Target deployment.

```yaml
apiVersion: sources.triggermesh.io/v1alpha1
kind: <Sample Source>
metadata:
  name: <Sample Source Name>
spec:
  sampleToken:
    secretKeyRef:
      name: <sample>
      key: <sample key>
  sink:
    ref:
      apiVersion: targets.triggermesh.io/v1alpha1
      kind: KafkaTarget
      name: triggermesh-kafka
```

### Sending messages to the Kafka Target

 A Kafka Target will, by default, accept any CloudEvent and pass the entire event into a message body.

> `Curl` can be used from a container in the cluster pointing to the ConfluentTarget exposed URL:

Reviewer: ```suggestion
`curl` can be used from a container in the cluster pointing to the `ConfluentTarget` exposed URL:
```

```

### Creating a Kafka Target with Kerberos-SSL

This section demonstrates how to configure a KafkaTarget to use Kerberos-SSL authentication.

Before creating the `KafkaTarget`, we are going to create some secrets that the `KafkaTarget` will need for the authentication with Kerberos + SSL.

- The kerberos config file.
- The kerberos keytab file.
- The CA Cert file.

```console
kubectl create secret generic config --from-file=krb5.conf
kubectl create secret generic keytab --from-file=krb5.keytab
kubectl create secret generic cacert --from-file=ca-cert.pem
```

```yaml
apiVersion: targets.triggermesh.io/v1alpha1
kind: KafkaTarget
metadata:
  name: sample
spec:
  salsEnable: true
  tlsEnable: true
  bootstrapServers:
  - kafka.example.com:9093
  topic: test-topic
  securityMechanism: GSSAPI
  kerberosAuth:
    username: kafka
    kerberosRealm: EXAMPLE.COM
    kerberosServiceName: kafka
    kerberosConfig:
      valueFromSecret:
        name: config
        key: krb5.conf
    kerberosKeytab:
      valueFromSecret:
        name: keytab
        key: krb5.keytab
  sslAuth:
    insecureSkipVerify: true
    sslCA:
      valueFromSecret:
        name: cacert
        key: ca-cert
```

> The following *ARE NOT* optional and without them the adapter will not deploy:

Reviewer: ```suggestion
 In order to configure the adapter correctly the following fields are mandatory:
```

- boostrapservers
- topic
- salsEnable
- tlsEnable

## Status

KafkaTarget requires secrets to be provided for the credentials. Once they are present it will create a Knative Service. Controller
logs and events can provide detailed information about the process. A Status
summary is added to the KafkaTarget object informing of the all conditions
that the target needs.

When ready the `status.address.url` will point to the internal point where Cloud Events should be sent.

```console
kubectl get kafkatarget
NAME     URL                                                   READY   REASON   AGE
sample   http://kafkatarget-sample.default.svc.cluster.local   True             1m
```

### Kafka Target as an event Sink

> A Kafka Target is addressable, which means it can be used as a Sink for Knative components.

Reviewer: ```suggestion
A `KafkaTarget` is addressable, which means it can be used as a Sink for Knative components.
```

```yaml
apiVersion: eventing.knative.dev/v1beta1
kind: Trigger
metadata:
  name: kafka-sample-trigger
spec:
  broker: default
  subscriber:
    ref:
      apiVersion: targets.triggermesh.io/v1alpha1
      kind: KafkaTarget
      name: triggermesh-kafka
```

> - A sample sink binding to a Kafka Target deployment.

Reviewer: ```suggestion
- A sample sink binding to a `KafkaTarget` deployment.
```

```yaml
apiVersion: sources.triggermesh.io/v1alpha1
kind: <Sample Source>
metadata:
  name: <Sample Source Name>
spec:
  sampleToken:
    secretKeyRef:
      name: <sample>
      key: <sample key>
  sink:
    ref:
      apiVersion: targets.triggermesh.io/v1alpha1
      kind: KafkaTarget
      name: triggermesh-kafka
```

### Sending messages to the Kafka Target

>  A Kafka Target will, by default, accept any CloudEvent and pass the entire event into a message body.

Reviewer: I'd recommend using KafkaTarget instead of Kafka Target everywhere
